st2api/st2api/controllers/v1/auth.py

```diff
@@ -0,0 +1,156 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pecan
+import six
+
+from pecan import abort
+from pecan.rest import RestController
+from mongoengine import ValidationError
+
+from st2common import log as logging
+from st2common.models.api.auth import ApiKeyAPI
+from st2common.models.api.base import jsexpose
+from st2common.exceptions.auth import ApiKeyNotFoundError
+from st2common.persistence.auth import ApiKey
+from st2common.rbac.types import PermissionType
+from st2common.rbac.decorators import request_user_has_permission
+from st2common.rbac.decorators import request_user_has_resource_permission
+from st2common.util import auth as auth_util
+
+http_client = six.moves.http_client
+
+LOG = logging.getLogger(__name__)
+
+__all__ = [
+    'ApiKeyController'
+]
+
+
+class ApiKeyController(RestController):
+    """
+    Implements the REST endpoint for managing the key value store.
+    """
+
+    def __init__(self):
+        super(ApiKeyController, self).__init__()
+        self.get_one_db_method = ApiKey.get_by_key_or_id
+
+    @request_user_has_resource_permission(permission_type=PermissionType.API_KEY_VIEW)
+    @jsexpose(arg_types=[str])
+    def get_one(self, api_key_id_or_key):
+        """
+            List api keys.
+
+            Handle:
+                GET /apikeys/1
+        """
+        api_key_db = None
+        try:
+            api_key_db = ApiKey.get_by_key_or_id(api_key_id_or_key)
+        except ApiKeyNotFoundError:
+            msg = 'ApiKey matching %s for reference and id not found.', api_key_id_or_key
+            LOG.exception(msg)
+            abort(http_client.NOT_FOUND, msg)
+
+        try:
+            return ApiKeyAPI.from_model(api_key_db)
+        except (ValidationError, ValueError) as e:
+            LOG.exception('Failed to serialize API key.')
+            abort(http_client.INTERNAL_SERVER_ERROR, str(e))
+
+    @request_user_has_permission(permission_type=PermissionType.API_KEY_VIEW)
```

Comment:
My experience is with the likes of google, twitter, slack etc where I can always go back to keys. Our users will probably prefer the view once approach if others are already using that approach. I will not immediately execute on this; will figure it out in this PR but after I get the validation part working. This also gives other some time to chime-in and voice opinion on this stuff :)

btw, a related question is do you feel we need to keep a one way hash of the key in the DB? Treat it like a password. This way anyone with access to the DB also cannot really make sense of the key or use it to abuse. It may be related to your concerns.